gradle 3.5

seanjreilly opened the pull request:
Created with `brew bump-formula-pr`.

neutric commented on Formula/gradle.rb:
```diff
@@ -1,8 +1,8 @@
   homepage "https://www.gradle.org/"
-  url "https://downloads.gradle.org/distributions/gradle-3.4.1-all.zip"
-  sha256 "ed7e9c8bb41bd10d4c9339c95b2f8b122f5bf13188bd90504a26e0f00b123b0d"
+  url "https://services.gradle.org/distributions/gradle-3.5-bin.zip"
```
Please don't switch from complete to binary-only distribution, the `with-all` option will thank you for it.

seanjreilly commented:
@neutric I have fixed this. However I have a few comments:

1. There are some drawbacks to selectively unpacking the zip file: it requires binary-only users to download more than they need and throw it away, which isn't being a good citizen to gradle's servers or users' bandwidth. It's also fragile because it requires internal knowledge of gradle's zip file layout, which could change. Another way to do this might be to change the url and hash from bin to all if the `with-all` flag is selected.

2. Far more importantly, an incorrect url that breaks the with-all option should have caused a test to fail. This didn't happen as far as I can see.

Anyway, fixed for now, but can I suggest that we address those deficiencies (at least the second one) in a separate PR? I'm willing to help if there's consensus on how to fix, but I admit that I could use some advice on good style wrt the tests as I haven't written many homebrew tests.

neutric commented:
Thanks for the fix and accompanying comments!

> There are some drawbacks to selectively unpacking the zip file: it requires binary-only users to download more than they need and throw it away, which isn't being a good citizen to gradle's servers or users' bandwidth. It's also fragile because it requires internal knowledge of gradle's zip file layout, which could change. Another way to do this might be to change the url and hash from bin to all if the with-all flag is selected.

The additional bandwidth used is a valid complaint, I'll look into the stats later to see whether it matters in this case. However, I am not sure whether we change url/hash based on which options are chosen. I'll wait for a more experience maintainer to chime in.

> Far more importantly, an incorrect url that breaks the with-all option should have caused a test to fail.

We test the default build, the rest is down to users reporting when things blow up. This is mainly due to limited CI resources, I think.

seanjreilly commented:
> We test the default build, the rest is down to users reporting when things blow up. This is mainly due to limited CI resources, I think.

@neutric  Assuming the url and hash aren't change for with-all, could we do something like add a test that checks that `-all` is in the url, and `-bin` isn't? It's know it's really hokey, but it would save you from having to manually inspect the URLs on every commit without having to perform a complete second installation...

neutric commented:
That feels like a kludge to me and is fragile, exactly what you, rightfully so, criticized above. What if upstream decides to change their URL pattern? Then we have to go change the test, which would not be part of the formula and thus inherently more cumbersome. It is really not a big deal, imho. Anything but a version number change warrants special attention and easily sticks out in the diff GitHub presents.

JCount commented on Formula/gradle.rb:
```diff
@@ -1,8 +1,8 @@
   homepage "https://www.gradle.org/"
-  url "https://downloads.gradle.org/distributions/gradle-3.4.1-all.zip"
-  sha256 "ed7e9c8bb41bd10d4c9339c95b2f8b122f5bf13188bd90504a26e0f00b123b0d"
+  url "https://services.gradle.org/distributions/gradle-3.5-all.zip"
```
You could leave this as `"https://downloads.gradle.org/distributions/gradle-3.5-all.zip"` since the `services.gradle.org` url is actually a 301 redirect.

seanjreilly replied:
Shouldn't we use the URL gradle advertises, in case they change what it redirects to? URLs are contracts, after all, and the gradle wrapper generates services.gradle.org urls.

neutric commented:
@seanjreilly `with-all` has some usage, so retiring it wouldn't be a good option. Changing the URL based on the selected option isn't something we generally do. There might be corner cases, but I don't know of one.

seanjreilly commented:
> Changing the URL based on the selected option isn't something we generally do.

@neutric: I don't know if it's done anymore, but the gradle recipe used to do exactly that with the `--devel` option and release candidate builds. Here's an example: https://github.com/Homebrew/homebrew-core/pull/3118/commits/58429bf861b939b31eaa5c188feb20c899b9d20f

neutric commented:
We have `--devel` and `--head`, yes, but they are not the same as `option`s.

seanjreilly commented:
Fair enough if that's something that's not normally done. I just remembered the `--devel` setting and figured that it _could_ work, but if that's not something people feel comfortable with, I've no desire to push it on anybody.

Anyway, this has all resulted in a lot more back-and-forth than a two-line change should have, and certainly more than previous gradle version update PRs have ever gotten. Sorry about that.

ilovezfs commented:
@seanjreilly thanks for the PR!
